codeforamerica/vita-min
# [FYST-1089] refactor twilio service

## Link to pivotal/JIRA issue
[FYST-1089](https://codeforamerica.atlassian.net/browse/FYST-1089)

## Is PM acceptance required? (delete one)
Yes - don't merge until JIRA issue is accepted!
**Reminder**: merge main into this branch and get green tests before merging to main

## What was done?
### Updates credentials file structure
The structure below would appear in each environment's credentials file. ~This structure reflects a simplified Twilio organization that isn't reflective of the current setup;~ restructuring the Twilio services and setup is a blocking dependency of this refactor (completed 11 Nov 24).

```
twilio:
  voice_phone_number: <i think there's just the one>
  gyr:
    account_sid: <value>
    auth_token: <value>
    messaging_service_id: <value>
  statefile:
    account_sid: <value>
    auth_token: <value>
    messaging_service_id: <value>
  ctc:
    account_sid: <value>
    auth_token: <value>
    messaging_service_id: <value>
```

### Updates `TwilioService` API
Enables vita-min to interact with many Twilio messaging services. The responsibility of determining which is positioned in the MultiTenantService. Now, to interact with Twilio using their REST api, you'll need to create an instance of `TwilioService` to represent a specific vita-min tenant, i.e., `TwilioService.new(:statefile)`. The returned instance knows how to do all the things the prior singleton version did, just scoped to the appropriate messaging service (one per tenant).

### Updates all the callsites and their specs
The bulk of the PR is updating all the places we interact with Twilio to use this updated interface. A bunch of related specs needed tweaking too.

## How to test?
Assuming all specs pass, a full test involves getting this into an environment with a valid Twilio setup (staging, demo, or production). Once there, the regular Twilio interactions (auth codes, document transfer, chats) should work as they always have. _New_ Twilio functionality (like FYST support) will come soon™️, in a separate PR/testing effort. 


[FYST-1089]: https://codeforamerica.atlassian.net/browse/FYST-1089?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
